Cross-Import Overlays

beccadax opened the pull request:
This PR adds cross-import overlays to Swift. Cross-import overlays allow a module to declare that, when it is imported in the same file as a second module, one or more additional modules should be imported as well. It can be used, for instance, to automatically import test helpers when your module is imported along with `XCTest`.

This implementation intentionally deviates from [the pitch thread](https://forums.swift.org/t/cross-import-overlays/30868) in certain ways:

* We use YAML, rather than JSON, to specify the cross-import information. YAML was selected because, unlike JSON, it allows comments. Their file extension is now `.swiftoverlay`.
* Rather than storing overlays in `.../Foo.swiftmodule/Overlays`, we store them in `.../Foo.swiftcrossimport`. This allows "thin" (i.e. non-platform-specific) Swift modules to declare cross-import overlays.
* Clang modules can now declare cross-import overlays by placing a `Foo.swiftcrossimport` directory next to the `module.modulemap` file that defines them.
* A mechanism has been added to declare platform-specific (rather, module-target-triple-specific) cross-import overlays.
* Some edge cases are handled differently from the pitch; for instance, we don't emit a diagnostic if both modules declare the overlay.

Other differences are either unintentional, incomplete, or undecided.

Fixes rdar://problem/59444602.

brentdax commented:
This is still early days. Still to do in this PR:

- [ ] Finalize filenames (I might change both .swiftoverlays and .yaml to .swiftoverlay)
- [ ] Clean up names of AST accessors
- [ ] Handle redundant imports better
- [ ] Diagnose or prevent cycles and other invalid situations, and test that behavior
- [ ] Decide whether re-exported modules should count for cross-import overlay loading, and implement that decision
- [ ] Add flag to disable the feature (and perhaps disable it by default so it can be merged as an experimental feature)

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please test

futurejones commented:
Can one of the admins verify this patch?

brentdax commented:
Feature-complete. To do:

- [ ] Rebase
- [ ] Self-review
- [ ] Clean up version history
- [ ] CI testing
- [ ] Peer review

brentdax commented:
@swift-ci test

brentdax commented:
@swift-ci test Windows platform

brentdax commented:
@swift-ci test

brentdax commented:
@swift-ci please test Linux platform

brentdax commented on include/swift/AST/Attr.h:
```diff
@@ -2103,6 +2103,15 @@ class DeclAttributes {
   }
 
+  /// Retrieve the first attribute with the given kind.
+  DeclAttribute *getAttribute(DeclAttrKind DK,
```
Did this end up as dead code?

brentdax replied:
No.

brentdax commented on include/swift/AST/Identifier.h:
```diff
@@ -79,6 +79,9 @@ class Identifier {
   explicit Identifier() : Pointer(nullptr) {}
+
+  // Allow instantiation by nullptr even though we don't allow other pointers.
+  Identifier(std::nullptr_t nul) : Identifier() {}
```
Check if I need this.

brentdax replied:
I don't. Deleted.

brentdax commented on include/swift/AST/SourceFile.h:
```diff
@@ -69,7 +72,8 @@ class SourceFile final : public FileUnit {
       assert(!(importOptions.contains(ImportFlags::Exported) &&
-               importOptions.contains(ImportFlags::ImplementationOnly)));
+               importOptions.contains(ImportFlags::ImplementationOnly)) ||
+             importOptions.contains(ImportFlags::Reserved));
```
Fix indentation.

brentdax replied:
It's actually correct—that condition is outside the parens.

brentdax commented on include/swift/AST/SourceFile.h:
```diff
@@ -514,4 +554,58 @@ inline void simple_display(llvm::raw_ostream &out, const SourceFile *SF) {
+struct DenseMapInfo<swift::SourceFile::ImportOptions> {
+  using ImportOptions = swift::SourceFile::ImportOptions;
+
+  using UnsignedDMI = DenseMapInfo<unsigned>;
```
Unsigned doesn’t match the underlying type’s size—will that cause problems with truncation?

brentdax replied:
Switching to a one-byte type.

brentdax commented on lib/AST/ModuleNameLookup.cpp:
```diff
@@ -127,6 +127,18 @@ void ModuleNameLookup<LookupStrategy>::lookupInModule(
+  if (!overlays.empty()) {
+    // If so, look in each of those overlays.
+    for (auto overlay : overlays)
+      lookupInModule(decls, overlay, accessPath, moduleScopeContext);
```
Will this work with multiple overlays on a single underlying module? Need a test for this.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -40,31 +43,250 @@ using namespace swift;
-namespace {  
-  class NameBinder {    
-  public:
+struct UnboundImport {
```
This should be in the anonymous namespace.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -40,31 +43,250 @@ using namespace swift;
+  NullablePtr<ModuleDecl> underlyingModule;
+
+  /// Create an UnboundImport from an ImportDecl.
+  explicit UnboundImport(ImportDecl *ID)
```
There’s probably enough logic in these initializers to justify pulling them further down the file.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -40,31 +43,250 @@ using namespace swift;
+};
+
+namespace {
+  struct ScopedImport {
```
Should this just be folded into UnboundImport?

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -40,31 +43,250 @@ using namespace swift;
+      if (auto scopedImport = std::get<1>(result))
+        unvalidatedScopedImports.emplace_back(ID, *scopedImport);
+
+      while(!unboundImports.empty())
```
Could use some comments here to explain what’s going on.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -40,31 +43,250 @@ using namespace swift;
+    { }
 
-    NameBinder(SourceFile &SF) : SF(SF), Context(SF.getASTContext()) {}
+    void visitImportDecl(ImportDecl *ID) {
```
Pull this function body downfile.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -158,236 +380,417 @@ static const char *getImportKindString(ImportKind kind) {
-    }
-    return;
-  }
+  ASTContext &Context = SF.getASTContext();
```
Capitalizing this didn’t save any code later in the function from being removed and re-added.

brentdax replied:
Changed.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -158,236 +380,417 @@ static const char *getImportKindString(ImportKind kind) {
+  if (M)
+    return true;
+
+  ASTContext &Context = SF.getASTContext();
```
Caps.

brentdax replied:
Changed.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -158,236 +380,417 @@ static const char *getImportKindString(ImportKind kind) {
+  return Context.AllocateCopy(decls);
+}
+
+NameBinder::BoundImport NameBinder::bindImport(UnboundImport &&I) {
```
I should reorder this code for clarity.

brentdax replied:
Done.

brentdax commented on test/NameBinding/Inputs/rewrite-module-triples.py:
```diff
@@ -0,0 +1,49 @@
+import shutil
+import sys
+
+# Does an in-place update of a flat Swift module to make it
```
Comment is out of date.

brentdax replied:
Fixed.

brentdax commented on test/NameBinding/import-cross-import-loading.swift:
```diff
@@ -0,0 +1,128 @@
+
+
+
+// TODO: check "-emit-loaded-module-trace-path -"
```
Should do this, or at least file a followup task to do it.

brentdax replied:
Done.

brentdax commented on test/NameBinding/import-transitive-cross-imports.swift:
```diff
@@ -0,0 +1,23 @@
+// RUN: %target-typecheck-verify-swift -enable-cross-import-overlays -I %t/include -I %t/lib/swift -F %t/Frameworks -DDIRECT_SECOND
+
+#if DIRECT_FIRST
+import ThinLibrary
```
I should use DeclaringModule and BystandingModule in this test instead—they’re more “real”.

brentdax replied:
No, it works better this way—the AlwaysImportedLibrary setup has a way to run negative tests.

brentdax commented on test/NameBinding/import-without-cross-imports.swift:
```diff
@@ -0,0 +1,78 @@
+
+//
+// The declaring module's import is scoped and that scope excludes
+// fromAlwaysImportedOverlay().
```
Fix this name.

brentdax replied:
Done.

brentdax commented on lib/AST/ModuleLoader.cpp:
```diff
@@ -54,4 +58,66 @@ DependencyTracker::getClangCollector() {
+
+  if (error && error != std::errc::no_such_file_or_directory) {
+    // FIXME: diagnose
+    llvm::report_fatal_error(error.message());
```
Diagnose this properly.

brentdax commented on lib/Sema/NameBinding.cpp:
```diff
@@ -158,236 +380,417 @@ static const char *getImportKindString(ImportKind kind) {
-                                        const SourceFile &SF) {
+static bool shouldImportSelfImportClang(bool isExported, const SourceFile &SF) {
   // FIXME: We use '@_exported' for fake overlays in testing.
-  if (ID->isExported())
```
I should remove the confusing, outdated comment above this line.

brentdax commented on test/NameBinding/import-cross-import-loading.swift:
```diff
@@ -0,0 +1,128 @@
+
+#if CLANG_FRAMEWORK
+import ClangFramework
+#endif
```
I should have test variants with a clang overlay—one where the clang module declares the cross-import overlays, another where the clang overlay does.

brentdax replied:
Done. (Using one test where both the Swift overlay and clang module declare a cross-import overlay.)

beccadax committed:
[NFC] Heavily refactor NameBinding

This commit refactors NameBinding.cpp virtually to the point of a rewrite. The goal is to make the import handling code process a struct containing information extracted from an ImportDecl, rather than the ImportDecl itself, so that a future commit can perform the same processing on imports that are not directly represented by an ImportDecl. The result takes more code, but it disentangles a knot of complicated logic into separate threads.

One semi-functional change is that validation of scoped imports (i.e. checking that the declaration actually exists) is now deferred until all of the file’s imports have been processed. Once cross-imports are supported, this will be necessary because scoped imports can select declarations from cross-import overlays, and the full set of cross-import overlays can’t be known until all imports have been seen.

Some comments refer to things that won’t exist until the next commit, like the visibleModules property. I’ll let you in on a litlte secret: I didn’t really do this all in one go.

beccadax committed:
Add frontend flags for staging cross-import overlays

beccadax committed:
Add swiftcrossimport and swiftoverlays file_types

beccadax committed:
Add support for loading cross-import files

beccadax committed:
Add mechanism for per-SourceFile overlays

When a “separately imported overlay” is added to a SourceFile, two things happen:

1. The direct import of the underlying module is removed from getImports*() by default. It is only visible if the caller passes ImportFilterKind:: ShadowedBySeparateOverlay. This means that non-module-scoped lookups will search _OverlayModule before searching its re-export UnderlyingModule, allowing it to shadow underlying declarations.

2. When you ask for lookupInModule() to look in the underlying module in that source file, it looks in the overlays instead. This means that UnderlyingModule.foo() can find declarations in _OverlayModule.

beccadax committed:
[NFC] Allow ImportedModuleDesc to be a DenseMap key

beccadax committed:
Load cross-import overlays

beccadax committed:
Test cross-import overlays

brentdax commented:
@swift-ci test

beccadax committed:
[NFC] Create one default value for EnableCrossImportOverlays

brentdax commented:
macOS passes tests. All tests fail on Linux due to failures to load modules. I suspect there's a systematic problem in the tests (case sensitivity?) rather than an actual bug in the implementation, but I'm investigating.

beccadax committed:
Correct for difference between BSD and GNU cp

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please test OS X platform

brentdax commented:
@swift-ci please smoke test Windows platform

brentdax commented:
Tagging for review:

* @varungandhi-apple, who pitched this and has a good idea of what it's for
* @hamishknight, who's touched NameBinding very recently
* @CodaFi, who will tell me exactly how cursed this whole thing is

hamishknight commented on lib/AST/DeclContext.cpp:
```diff
@@ -287,6 +287,12 @@ DeclContext *DeclContext::getModuleScopeContext() const {
+void DeclContext::getSeparatelyImportedOverlays(
+    ModuleDecl *declaring, SmallVectorImpl<ModuleDecl *> &overlays) const {
+  if (auto SF = getParentSourceFile())
+    SF->getSeparatelyImportedOverlays(declaring, overlays);
```
Do we want to pick up overlays introduced by `@_exported` cross-imports in other source files?

brentdax replied:
I think we do. Consider the case where you import, say, UIKit and MyModule, and there's a cross-import available for MyModule and Foundation. Shouldn't we import it? You would certainly expect to be able to call those APIs.

hamishknight replied:
@brentdax Sorry, what I mean is that if you have for example a cross-import for module `X` and module `Y`, and in a file `b.swift` you have:

```swift
@_exported import X
@_exported import Y
```

Then shouldn't another file in the same module as `b.swift`, say `a.swift`, also pickup the cross-import overlay? As far as I can tell, the current implementation will only check for cross-imports from `a.swift`'s imports.

brentdax replied:
This is a very, very good catch. I don't want to do even more work without merging, so I'll fix this in a follow-up PR.

brentdax replied:
I've filed a radar on myself for this and added a comment describing the issue.

hamishknight commented on lib/Sema/NameBinding.cpp:
```diff
@@ -158,282 +673,239 @@ static const char *getImportKindString(ImportKind kind) {
+  // Per getTopLevelModule(), topLevelModule should only be nullptr if we are
+  // importing a submodule, and we should never do a scoped import of a
+  // submodule.
+//  assert(topLevelModule && "scoped import of a submodule import?");
```
Should this assertion be uncommented?

brentdax replied:
Oh, I forgot about this. I refactored the code and it no longer made sense to express this invariant here. I'll fix it.

brentdax replied:
This invariant sort of went away when `topLevelModule` became a local in `NameBinder::bindImport()`, but I've added a related one to the `BoundImport` constructor.

hamishknight reviewed (approved):
Looks great! Just a couple of questions

varungandhi-apple commented:
> Some edge cases are handled differently from the pitch; for instance, we don't emit a diagnostic if both modules declare the overlay.

What is the reason behind not emitting a diagnostic? Less implementation work? Seems like it creates an edge case that is potentially hard to debug.

varungandhi-apple commented on include/swift/Basic/FileTypes.def:
```diff
@@ -73,6 +73,13 @@ TYPE("module-trace",        ModuleTrace,               "trace.json",      "")
 TYPE("index-data",          IndexData,                 "",                "")
 TYPE("opt-record",          OptRecord,                 "opt.yaml",        "")
 
+// Overlay files declare wrapper modules, called "separately-imported overlays",
```
Could you add entries to Lexicon.rst for separately-imported overlay, cross-import overlay and maybe even cross-import dir?

brentdax replied:
That would be a good idea.

To respond to this from above:

> What is the reason behind not emitting a diagnostic?

Mainly staging—if you're trying to switch from one to the other and accidentally end up with both modules declaring the cross-import, should your clients all break? It would cause some pretty psychedelic name lookup behavior, though, so maybe we should emit a warning.

varungandhi-apple replied:
Yeah name lookup is the thing I'm worried about. If it is a warning, it is still ok (we can live with that) for a temporary thing. If it creates some kind of error due to funky name lookup and we don't have that warning, debugging the situation will not be fun (esp. if you are on vacation when we get the bug report 😝).

brentdax replied:
I'll add a warning, and make sure it makes it clear that this is the modules' fault.

brentdax replied:
Added.

varungandhi-apple commented on include/swift/AST/Module.h:
```diff
@@ -203,6 +205,10 @@ class ModuleDecl : public DeclContext, public TypeDecl {
 
   SmallVector<FileUnit *, 2> Files;
 
+  friend OverlayFile;
```
Is there a specific reason you separated out `class OverlayFile;` and `friend OverlayFile;` instead of having one `friend class OverlayFile;`? [I usually prefer the latter, I don't know if there is a reason to prefer the former.]

varungandhi-apple replied:
Also, I can't quite tell where this `friend` relation is being used -- could you point that out?

brentdax replied:
The need for friend might have been refactored away—I’m checking this.

brentdax replied:
Yeah, I don't need it anymore. Good catch!

brentdax replied:
Removed.

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+
+  auto bufOrError = fs.getBufferForFile(filePath);
+  if (!bufOrError) {
+    M->getASTContext().Diags
```
We already have `auto &ctx = M->getASTContext()` earlier so we can reuse the local variable here and a bit below.

brentdax replied:
Fixed.

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+  ///
+  /// If empty, \c overlayModuleNames is valid and will contain the list of
+  /// module names unless loading the file failed.
+  StringRef filePath;
```
I don't quite understand the logic here...

1. Why do we have an empty path when loading succeeds?
2. If loading the file fails, what state is this struct in? Is the path supposed to be empty and the vector is also empty?

---

There are three states we want to represent:

1. We have not yet read the file - Having the path here is necessary.
2. We tried reading the file and failed (I/O, deserialization etc.) - Having the path here is useful for debugging.
3. We tried reading the file and succeeded - Having the path here is potentially useful for debugging, in case we have trouble with loading the modules later on.

I think it would be simpler to have a dedicated 3-state enum instead of trying to squeeze the states into empty/non-empty combinations.

varungandhi-apple replied:
This would also simplify the assertions in different places. We could have a single function (say `isValid`) which does the following checks:

1. The path is always non-empty.
2. The vector is empty if we have not yet read the file and non-empty if we did read the file successfully.

and use that check consistently.

brentdax replied:
Okay, let's start with the preferred model. In Swift, I would probably define this type like this:

```swift
enum OverlayFile {
  case unloaded(filePath: String)
  case loaded(overlayNames: Set<Identifier>)
}
```

Could I add a third case for `cannotLoad`? Sure, but why would I? Once the error has been diagnosed, the rest of the compiler does not need to reason about the `OverlayFile.cannotLoad` case—it will behave exactly as though it were `OverlayFile.loaded(overlayNames: [])`. So why add this third case and make the rest of the code accommodate it?

-----

With that aside: the Swift model is a sum type, but I've actually implemented it with a product type. This is entirely because it would be a pain in the rear to write it as a sum type in our dialect of C++. We don't have `std::variant` in C++ 14. The values are not compatible with `llvm::PointerUnion`.

(Actually, it looks like we just added a TaggedUnion.h that could make this reasonable. Maybe I should just use that.)

varungandhi-apple replied:
I agree with you that the minimal model that satisfies the requirements is the one you pointed out. The issue with the minimal model IMO is that it potentially makes debugging harder -- we are forgetting information (dropping the path, not keeping track of whether there was a load failure) even though it doesn't seem necessary to do that. It would be a different equation if we have thousands or hundreds of thousands of these, which is true for types, where we try to squeeze information into a low memory footprint.

If later down the line, say, we see a problem that we expected one cross-import overlay to be loaded but something else got loaded, debugging that forces us to look at the state in between (and how it is changing) when we have partial information instead of having the option to look at it in the end when we have maximum information.

---

> Once the error has been diagnosed, the rest of the compiler does not need to reason about the OverlayFile.cannotLoad case—it will behave exactly as though it were OverlayFile.loaded(overlayNames: []). So why add this third case and make the rest of the code accommodate it?

This is a good point. I was under the impression that not a lot of code was touching this directly but I haven't looked at the later commits (at 3/10 atm). If that's the case, do you think it might be useful to separate it out into two types an `OverlayFileWIP` and an `OverlayFile` with the latter only having the success state?

brentdax replied:
I tried going full `TaggedUnion` to model this properly, but ran into some weird snarl involving Identifier's move constructor. I'll add a tristate.

brentdax replied:
```
  enum class State { pending, loaded, failed };
  State state = State::pending;
```

varungandhi-apple replied:
I've made a small comment about the naming convention `Pending` etc. (instead of `pending` etc.) but looks good now. 👍

varungandhi-apple commented on lib/AST/ModuleLoader.cpp:
```diff
@@ -54,4 +59,73 @@ DependencyTracker::getClangCollector() {
+    return;
+  path::remove_filename(dirPath);
+
+  // <parent-dir>/<module-name>.swiftcrossimport
```
I can't quite remember -- what was the reason behind us deciding to support both `swiftcrossimport` as a file and as a directory? Avoid more disk usage in case all the architecture-specific files are the same?

brentdax replied:
The comments may not be clear enough here. What's happening is that:

1. `getModuleDefiningPath()` returns the path of the `module.map` file or `Foo.swiftmodule` directory.
2. We strip away the last component to get the directory that file is in.
3. We append `Foo.swiftcrossimport` to it, then search that directory.
4. We append the target's module triple to it, then search that directory.
5. If there's a target variant (for Catalyst), we remove the target's module triple and append the target variant's module triple, then search that directory.

If this isn't being conveyed well, how would you clarify it?

varungandhi-apple replied:
Ah, gotcha'. Those are all directory searches. I just misread the first one.

I think the code is 👍 but I'm wondering if there should be extra context. Like, when would we use the top-level directory over the architecture-specific directory? Is it a question of not unnecessarily duplicating files?

Also, having a line like `// E.g. for Catalyst` would be useful immediately above the `// <parent-dir>/<module-name>.swiftcrossimport/<targetvariant-module-triple>`.

brentdax replied:
Honestly, I don't know if we'll end up needing the target-specific directories; I added them because (a) it's easy to just do it now and (b) if we do need them, we will inevitably discover this only at the eleventh hour when there might be pressure to insert a hack instead.

brentdax replied:
Expanded comments.

varungandhi-apple commented on lib/AST/ModuleLoader.cpp:
```diff
@@ -54,4 +59,73 @@ DependencyTracker::getClangCollector() {
+
+  auto &langOpts = module->getASTContext().LangOpts;
+
+  SmallString<64> dirPath{file->getModuleDefiningPath()};
```
I didn't realize you could use braces here instead of parens. Huh.

brentdax replied:
That was added to C++ to deal with the [most vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse).

varungandhi-apple commented on include/swift/AST/ModuleLoader.h:
```diff
@@ -154,6 +155,10 @@ class ModuleLoader {
+
+  /// Discover overlays declared alongside this file and add infomation about
+  /// them to it.
+  void findOverlayFiles(SourceLoc diagLoc, ModuleDecl *module, FileUnit *file);
```
This seems like a good place to use `swift/Basic/Located` -- what do you think? The `diagLoc` corresponds to the import of the `ModuleDecl`, right?

brentdax replied:
`diagLoc` is the location where the `ModuleDecl` was imported, but it's not the location of the `ModuleDecl` itself. If this were an `ImportDecl` (and that class didn't have embedded locations), or if it were an `Identifier moduleName` along with the source location of the identifier, I would agree with you.

varungandhi-apple replied:
Gotcha', makes sense.

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+  /// to not be loaded again.
+  ///
+  /// The result can be empty, either because of an error or because the file
+  /// didn't contain any overlay module names.
```
I don't understand this bit. If the file didn't contain any overlay module names, is that not an error?

brentdax replied:
I don't diagnose it as an error—it's just an array with zero entries. Maybe I should.

varungandhi-apple replied:
That seems like a mistake on the part of whoever wrote the overlay. Since we will see the problem in a client, not the overlay author... I'm feeling a bit divided on if we should give a hard error or not. It seems unlikely that someone would ship a totally broken overlay with no entries so maybe the hard error is fine, and we want to detect that sooner rather than later in case that happens. A warning is definitely not what we want because the client cannot suppress the warning without going in and deleting/modifying the file.

brentdax replied:
Suppose someone, for instance, ran their swiftoverlay files through the C preprocessor (or gyb, or whatever) so they could use `#if` to remove overlays related to unreleased features:

```
---
version: 1
modules:
#if !HIDE_RAY_GUNS
- name: _MyModule_OtherModule_RayGuns
#endif
#if !HIDE_FLYING_CARS
- name: _MyModule_OtherModule_FlyingCars
#endif
```

This could totally result in a swiftoverlay file with no modules listed. It'd probably be better if they erased the file, but it's not really doing any harm.

varungandhi-apple replied:
That's a good point! I'm convinced.

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+      loadOverlayModuleNames(M, diagLoc, bystandingModule);
+      filePath = "";
+    }
+    return overlayModuleNames;
```
Is this guaranteed to perform a copy of the underlying array? My understanding is that we are providing an interior reference to the buffer which is a field...

brentdax replied:
`TinyPtrVector`s manage their own storage. This one is nested in an `OverlayFile`, and those are always allocated in the `ASTContext`. So I believe this array will remain valid for the lifetime of the AST context.

varungandhi-apple replied:
That makes sense.

varungandhi-apple commented on test/CrossImport/horrible.swift:
```diff
@@ -0,0 +1,44 @@
+// RUN: %empty-directory(%t)
+// RUN: cp -r %S/Inputs/lib-templates/ %t/
+
+// RUN: %target-typecheck-verify-swift -enable-cross-import-overlays -I %t/lib/swift
```
Could you modify this test case (or add another one, whichever is easier) to make sure cross-import overlays are working with the trace files and `.d` files?

brentdax replied:
CrossImport/loading.swift tests `-emit-dependencies-path`, and CrossImport/module-trace.swift tests `-emit-loaded-module-trace-path`.

varungandhi-apple replied:
Oops, those are in a later commit. Sorry!

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+                                SmallVectorImpl<Identifier> &overlayNames,
+                                SourceLoc diagLoc) {
+  if (getName() == bystanderName)
+    // We don't currently support self-cross-imports.
```
This seems like this should be a hard `llvm::report_fatal_error` instead of a silent pass, in case someone tries to do this [same as with the empty case], although I don't see a reason for _why_ they would want to do this.

brentdax replied:
`findDeclaredCrossImportOverlays()` is where other parts of the compiler can ask module X, "Do you have any cross-imports with module Y?" It is likely that, occasionally, callers will naïvely ask module X, "Do you have any cross-imports with module X?" If they do, emitting a fatal error seems kind of unfriendly—we should just quietly skip the check.

We could certainly also have a check for self-cross-imports (or self-overlaying) in the vicinity of `OverlayFile`, but that would be something different.

varungandhi-apple replied:
That makes sense. Let me get back to this tomorrow.

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+    // We don't currently support self-cross-imports.
+    return;
+
+  for (auto &crossImportFile : declaredCrossImports[bystanderName])
```
Do you think it would be useful to have an assertion here that `declaredCrossImports.find(bystanderName) != declaredCrossImports.end()`? IIUC, we wouldn't want to accidentally create a default value here.

brentdax replied:
The default empty value is fine, but maybe I should use `lookup` instead of the subscript to make sure we don't insert it.

brentdax replied:
Done.

beccadax committed:
Add a warning about redundant cross-import declarations

These are mostly harmless, except that they make the two module names synonymous in qualified lookup. A hard error seems too aggressive for something that could easily be caused by uncoordinated changes to two modules, so warn instead.

varungandhi-apple commented on lib/AST/Module.cpp:
```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
+
+  auto contents = std::move(*contentsOrErr);
+
+  for (const auto &module : contents.modules) {
```
I think this'll create an extra copy if we use `const auto &` instead of `auto &`, not 100% sure...

varungandhi-apple commented on lib/AST/ModuleLoader.cpp:
```diff
@@ -54,4 +59,73 @@ DependencyTracker::getClangCollector() {
+
+    module->addCrossImportOverlayFile(file);
+
+    // FIXME: Better to add it only if we load it.
```
Not asking you to do this right now, but any thoughts on how we might propagate sufficient information to track this in a fine-grained manner?

brentdax replied:
We'd need to thread `tracker` through to `OverlayFile` and have it only add the dependency when `loadOverlayModuleNames()` is called. Nothing difficult, just annoying.

varungandhi-apple commented on test/CrossImport/horrible.swift:
```diff
@@ -0,0 +1,44 @@
+// expected-error@-1 {{cannot list cross-import overlays for 'UnitaryGoose': Not a directory}}
+
+// FIXME: It might be better to diagnose these errors on HorribleGoose's import
+// decl, since they actually belong to HorribleGoose.
```
Could you file a radar to improve the diagnostic location for this one? I think for the same file case, this is probably fine, but for the `@_exported import` case [when we add that], it might be a bit of a head-scratcher what is going on if you don't understand the name lookup there.

brentdax replied:
The SourceLoc thing is actually kind of interesting. What I'm currently doing is diagnosing with the location of the `ImportDecl` that caused us to load the swiftoverlay file. This has the upside that it will always be in user code, even if it's been imported indirectly, and that commenting out that ImportDecl will always cause the error to either go away or move. But it has the downsides that it's not necessarily the import *most directly connected* to the overlay file, and that it's not necessarily clear why the compiler decided it needed that cross-import there.

I think it would be better to include notes pointing to all of the imports involved in a cross-import overlay (hopefully even using generated interfaces to point to `@_exported imports` in modules), but `ModuleDecl::ImportedModule` doesn't include the necessary information to do that, so I think it will be difficult. The current implementation reliably gives us a reasonably actionable source location without doing a lot of engineering.

But I can certainly file an SR if you'd like.

varungandhi-apple replied:
> I think it would be better to include notes pointing to all of the imports involved in a cross-import overlay (hopefully even using generated interfaces to point to @_exported imports in modules), but ModuleDecl::ImportedModule doesn't include the necessary information to do that, so I think it will be difficult. The current implementation reliably gives us a reasonably actionable source location without doing a lot of engineering.
>
> But I can certainly file an SR if you'd like.

I agree having that would be ideal. I think filing an SR would be useful even if we are not implementing this right away. It indicates that this is a known problem, in case someone runs into it.

brentdax replied:
https://bugs.swift.org/browse/SR-12223

I'll add that number to the comment, too.

brentdax replied:
Done.

varungandhi-apple commented on test/CrossImport/Inputs/lib-templates/lib/swift/HorribleGoose.swiftcrossimport/GibberishGoose.swiftoverlay:
```diff
@@ -0,0 +1 @@
+It's a lovely morning in the SDK, and you are a horrible goose.
```
😆😆😆

beccadax committed:
[NFC] Improvements suggested in code review

Thank you, @hamishknight and @varungandhi-apple.

beccadax committed:
Don’t count submodules when cross-importing

It turns out that, if you pull in any nontrivial module, there are thousands of submodules and none of them could possibly have a  cross-import overlay. Avoid evaluating them.

brentdax commented:
Ha, Slava cleaned something up beneath me and created a merge conflict. I have a rebase locally, but I don't want to force-push mid-review.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
```
It would be useful to document the difference between `modulePath` and `declPath`.

brentdax replied:
I've documented all of these members.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    /// This property should only be used to control where information is added
+    /// to the AST. Don't retrieve information directly from the ImportDecl; use
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
```
Huh, I didn't realize we had a utility class for this. Neat.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
```
Bit confused here -- why do we need both `DeclAttributes` and `ImportOptions`? Is `ImportOptions` a parsed/cached representation primarily present to avoid reparsing the attributes?

brentdax replied:
`options` is always set; `attrs` is only set on `UnboundImport`s which directly correspond to an `ImportDecl` (not ones for cross-import overlays). When we validate `options`, we will use `attrs` if available to provide fix-its removing the invalid attributes.

In other words, `options` is the source of truth for binding the import; `attrs` is used if available to emit better diagnostics.

I should explain this more explicitly.

brentdax replied:
I reworked this whole thing so that `attrs` is no longer a separate property and `ID` and `underlyingModule` are merged into a single `PointerUnion` (since we will have exactly one of the two at any given time).

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    explicit UnboundImport(ImportDecl *ID);
+
+    /// Make sure the import is not a self-import.
+    bool checkNotTautological(const SourceFile &SF);
```
It may not be obvious what "tautological" means in this context (e.g. one might think that you can't be allowed to import the same module because then you could write something like `@_implementationOnly import MyName` which is impossible.). Maybe `checkNotSelfImport`?

brentdax replied:
Some self-imports are allowed, so I prefer this terminology. I'll expand the doc comment to explain:

```
    /// Diagnoses if the import would simply load the module \c SF already
    /// belongs to, with no actual effect.
    ///
    /// Some apparent self-imports do actually load a different module; this
    /// method allows them.
```

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    return M;
+  }
+
+  if (topLevelModule == SF.getParentModule())
```
Wrt this check and the `return topLevelModule` below: at this point in the code, is it not true that `topLevelModule == nullptr`?

brentdax replied:
No—the previous condition is `if (!topLevelModule)`, so we only reach this code if `topLevelModule` is not null.

(`!` is a tiny operator, isn't it?)

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    // submodules.
+    assert(!M->getFiles().empty() &&
+           isa<ClangModuleUnit>(M->getFiles().front()));
+    return M;
```
Should this be:

```
    assert(!topLevelModule->getFiles().empty() &&
            isa<ClangModuleUnit>(topLevelModule->getFiles().front()));
     return topLevelModule;
```

or does `M == topLevelModule` here?

varungandhi-apple replied:
Oh nvm, this is resolved by the later one.

brentdax replied:
This function was [extracted and only lightly reformatted](https://github.com/apple/swift/blame/6616e812c6f508fc4ad4c39c7e0d977c9add289c/lib/Sema/NameBinding.cpp#L212-L230) from the previous implementation, so I'm not familiar with its full history. Looking at the blame, this particular bit was added in [this PR](https://github.com/apple/swift/pull/14673).

Basically, we used to just assert that every submodule has a top-level module. However, there's a corner case in clang where you can ask to import a submodule and get a top-level module instead. We handled that by asserting that, if the top-level module is nullptr, then the submodule is a clang module, and then returning the clang module as though it were the top-level module (which of course it actually is).

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    /// returns \p M. If it is a submodule, returns either the parent module of
+    /// \p M or \c nullptr if the current module is the parent (which can happen
+    /// in a mixed-language framework).
+    NullablePtr<ModuleDecl> getTopLevelModule(ModuleDecl *M, SourceFile &SF);
```
In the doc comment: "either the parent module" isn't quite right, I think, because the submodule may be nested and we will return the root of the hierarchy, not the immediate parent.

brentdax replied:
Right—I was using "parent module" too loosely. New doc comment:
```
    /// Find the top-level module for this module; that is, if \p M is the
    /// module \c Foo.Bar.Baz, this finds \c Foo.
    ///
    /// Specifically, this method returns:
    ///
    /// \li \p M if \p M is a top-level module.
    /// \li \c nullptr if \p M is a submodule of \c SF's parent module. (This
    ///     corner case can occur in mixed-source frameworks, where Swift code
    ///     can import a Clang submodule of itself.)
    /// \li The top-level parent (i.e. ancestor with no parent) module above
    ///     \p M otherwise.
```

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+
+BoundImport::BoundImport(UnboundImport unbound, ImportedModuleDesc desc,
+                         ModuleDecl *module, bool needsScopeValidation)
+  : unbound(unbound), desc(desc), module(module),
```
Is it ok for the `ModuleDecl *` to be null (e.g. in the case of mixed-source frameworks?)? If not, it would be good to add an assertion here.

brentdax replied:
I actually did this in response to one of Hamish's review comments.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+
+  /// Represents an import whose options have been checked and module has been
+  /// loaded, but its scope (if any) has not been validated.
+  struct BoundImport {
```
1. I think it is good that you separated out `BoundImport` and `UnboundImport`. This makes the separation of responsibilities clearer.
2. It is not entirely clear what "scope has not been validated" means here, and the doc comment for `validateScope` doesn't help. Moreover, there is a separate field `needsScopeValidation` defaults to `false`, which seems to contradict this (if it has not been validated, then why is the default value false?). Perhaps phrase it to something like:

```
  /// Represents an import whose options have been checked and module has been
   /// loaded, but its scope has potentially not been validated.
   ///
   /// Scope validation is needed for imports like `import func Mod.fn` where we check that `Mod`
   /// does contain `fn`.
```

?
```

brentdax replied:
```
  /// Represents an import whose options have been checked and module has been
  /// loaded, but its scope (if it's a scoped import) has not been validated
  /// and it has not been added to \c SF.
```

brentdax replied:
I've also expanded the doc comment on `validateScope`:

```
    /// Validate the scope of the import, if needed.
    ///
    /// A "scoped import" is an import which only covers one particular
    /// declaration, such as:
    ///
    ///     import class Foundation.NSString
    ///
    /// We validate the scope by making sure that the named declaration exists
    /// and is of the kind indicated by the keyword. This can't be done until we
    /// have bound all cross-import overlays, since a cross-import overlay could
    /// provide the declaration.
```

And I've flipped the sense of the `needsScopeValidation` so I could give it a better name and more accurate doc comment:

```
    /// If \c true, another, more specific \c BoundImport will validate the same
    /// scope as this import, so validating this one is not necessary.
    bool scopeValidatedElsewhere = true;
```

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
```
Nitpick: Field names should be begin with uppercase letters. (Similarly in `BoundImport` and `NameBinder`).

brentdax replied:
We've actually been transitioning away from this style in general. I made an exception for `ID` because `id` for an "import decl" just seems weird, but I'm thinking about changing this anyway.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+
+    /// Postprocess the imports this NameBinder has bound and collect them into
+    /// a vector.
+    void finishImports(SmallVectorImpl<ImportedModuleDesc> &imports);
```
Nitpick: "into a vector" is not necessary.

brentdax replied:
Changed to:
```
    /// Postprocess the imports this NameBinder has bound and collect them into
    /// \p imports.
```
which makes that bit actually useful.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    void validateScope(SourceFile &SF);
+  };
+
+  class NameBinder : public DeclVisitor<NameBinder> {
```
Mark `NameBinder` as final? I don't think anything is inheriting from it.

varungandhi-apple replied:
Do you think it would be useful to have a small explanation for the different stages imports go through? Something like (assuming I've understood the details correctly):

```
[Note: import-name-binding-stages]
Imports are roughly processed in three stages:

1. Unbound: The import has been synthesized but is not known which ModuleDecl it corresponds to. 
2. Bound: We know which ModuleDecl the import corresponds to, but we may not have validated it fully.
3. Bound and validated: The import is bound to a ModuleDecl and has been fully validated so we are free to use it.
```

This explanation can be referred to in the doc comments for `addImport` and `bindImport` (right now, their doc comments seems to be overly focused on describing the implementation, which seems dissatisfying) with a `see [Note: import-name-binding-stages]`.

[Feel free to skip this suggestion if you feel like rephrasing it correctly might take too much time.]

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+/// UnresolvedDeclRefExpr nodes for unresolved value names, and we may have
+/// unresolved type names as well. This handles import directives and forward
+/// references.
+void swift::performNameBinding(SourceFile &SF) {
```
I don't see a connection between the first sentence and the second sentence in the doc comment's paragraph

```
 /// At this point parsing has been performed, but we still have
 /// UnresolvedDeclRefExpr nodes for unresolved value names, and we may have
 /// unresolved type names as well. This handles import directives and forward
 /// references.
```

Are they related? If not, can you separate the second sentence into a separate paragraph? Something like:

```
 /// At the time this function is called, parsing has been performed, but we still have
 /// UnresolvedDeclRefExpr nodes for unresolved value names, and we may have
 /// unresolved type names as well. Those names are resolved.
 ///
 /// We also handle import directives and forward references here.
```

varungandhi-apple replied:
Oh looks like this was not part of your PR. Nvm, we can fix this later.

brentdax replied:
I think the first sentence is a dead comment. My impression is that, once upon a time, NameBinding.cpp bound *all* of the names; now it just binds a few that don't work through normal lookup.

brentdax replied:
```
/// performNameBinding - Once parsing is complete, this walks the AST to
/// resolve names and do other top-level validation.
///
/// Most names are actually bound by the type checker, but before we can
/// type-check a source file, we need to make declarations imported from other
/// modules available and build tables of the operators and precedecence groups
/// declared in that file. Name binding processes top-level \c ImportDecl,
/// \c OperatorDecl, and \c PrecedenceGroupDecl nodes to perform these tasks,
/// along with related validation.
///
/// Name binding operates on a parsed but otherwise unvalidated AST.
```

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+  class NameBinder : public DeclVisitor<NameBinder> {
     SourceFile &SF;
-    ASTContext &Context;
+    ASTContext &ctx;
```
Accidental rename?

brentdax replied:
No, intentional; I'm using `ctx` in new code and didn't want it to be inconsistent.

varungandhi-apple commented on lib/Sema/NameBinding.cpp:
```diff
@@ -36,60 +37,539 @@
+    // yourself--which doesn't make any sense to do. Shouldn't we diagnose that?
+    //
+    // I'm leaving this alone for now because I'm trying to refactor without
+    // changing behavior, but it smells funny.
```
Thanks for writing a detailed comment explaining the problem here!

varungandhi-apple commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -174,6 +174,11 @@ ERROR(cannot_load_swiftoverlay_file, none,
+        "modules %0 and %1 both declare module %2 as a cross-import overlay, "
+        "which may cause paradoxical behavior when looking up names in them; "
+        "please report this bug to the maintainers of these modules",
+        (Identifier, Identifier, Identifier))
```
I like this diagnostic. It is wordier than our typical diagnostics but it explains things clearly.

beccadax committed:
[NFC] Additional readability improvements

Courtesy of Varun.

beccadax committed:
[NFC] Replace three fields in UnboundImport with a PointerUnion

It’s not often that a PointerUnion makes the code *more* safe and clear, so let’s take it.

brentdax commented:
@swift-ci please test

brentdax commented:
@varungandhi-apple I'm going to merge this, but feel free to post additional comments on the last round of changes.

varungandhi-apple commented:
LGTM 😄

drodriguez commented:
Just a note: a couple of tests seem broken on Windows (started in https://ci-external.swift.org/job/oss-swift-windows-x86_64/2907/ and https://ci-external.swift.org/job/oss-swift-windows-x86_64-vs2019/103/). One of the failing tests seems to be simple error message differences in Windows. The other seems to be some path comparison that doesn't normalize paths.

/cc @compnerd

devincoughlin commented:
We'll look into this.

devincoughlin commented:
I have a PR up at https://github.com/apple/swift/pull/29962 to unblock windows testing.

varungandhi-apple commented:
Related PR: [just to get the GitHub cross-linking to show up] https://github.com/apple/swift/pull/29965

buttaface commented on test/CrossImport/skip-submodules.swift:
```diff
@@ -0,0 +1,16 @@
+// Make sure we don't consider submodules, because there are...a lot of them.
+
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -typecheck -enable-cross-import-overlays -Xllvm -debug-only=swift-name-binding %s 2>%t.txt
```
In addition, I believe this test should have a `REQUIRES: asserts` as it fails for me on Android when the tests are run without assertions:
```
swift (LLVM option parsing): Unknown command line argument '-debug-only=swift-name-binding'.  Try: 'swift (LLVM option parsing) --help'
swift (LLVM option parsing): Did you mean '--debug-pass=swift-name-binding'?
```
